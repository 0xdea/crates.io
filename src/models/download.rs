use crate::models::Version;
use crate::schema::version_downloads;
use chrono::NaiveDate;

#[derive(Queryable, Identifiable, Associations, Debug, Clone, Copy)]
#[diesel(belongs_to(Version))]
#[diesel(primary_key(version_id, date))]
pub struct VersionDownload {
    pub version_id: i32,
    pub downloads: i32,
    pub counted: i32,
    pub date: NaiveDate,
    pub processed: bool,
}
