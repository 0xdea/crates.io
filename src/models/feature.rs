use std::collections::BTreeMap;

pub type FeaturesMap = BTreeMap<String, Vec<String>>;

/// Splits the given [`FeaturesMap`] into two [`FeaturesMap`]s based on their
/// values.
///
/// See <https://rust-lang.github.io/rfcs/3143-cargo-weak-namespaced-features.html>.
pub fn split_features(features: FeaturesMap) -> (FeaturesMap, FeaturesMap) {
    features.into_iter().partition(|(_k, vals)| {
        !vals
            .iter()
            .any(|v| v.starts_with("dep:") || v.contains("?/"))
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_compact_debug_snapshot;

    #[test]
    fn test_split_features_no_deps() {
        let mut features = FeaturesMap::new();
        features.insert(
            "feature1".to_string(),
            vec!["val1".to_string(), "val2".to_string()],
        );
        features.insert("feature2".to_string(), vec!["val3".to_string()]);

        let (features, features2) = split_features(features);

        assert_compact_debug_snapshot!(features, @r#"{"feature1": ["val1", "val2"], "feature2": ["val3"]}"#);
        assert_compact_debug_snapshot!(features2, @"{}");
    }

    #[test]
    fn test_split_features_with_deps() {
        let mut features = FeaturesMap::new();
        features.insert(
            "feature1".to_string(),
            vec!["dep:val1".to_string(), "val2".to_string()],
        );
        features.insert(
            "feature2".to_string(),
            vec!["val3".to_string(), "val4?/val5".to_string()],
        );

        let (features, features2) = split_features(features);

        assert_compact_debug_snapshot!(features, @"{}");
        assert_compact_debug_snapshot!(features2, @r#"{"feature1": ["dep:val1", "val2"], "feature2": ["val3", "val4?/val5"]}"#);
    }

    #[test]
    fn test_split_features_mixed() {
        let mut features = FeaturesMap::new();
        features.insert(
            "feature1".to_string(),
            vec!["val1".to_string(), "val2".to_string()],
        );
        features.insert("feature2".to_string(), vec!["dep:val3".to_string()]);
        features.insert(
            "feature3".to_string(),
            vec!["val4".to_string(), "val5?/val6".to_string()],
        );

        let (features, features2) = split_features(features);

        assert_compact_debug_snapshot!(features, @r#"{"feature1": ["val1", "val2"]}"#);
        assert_compact_debug_snapshot!(features2, @r#"{"feature2": ["dep:val3"], "feature3": ["val4", "val5?/val6"]}"#);
    }

    #[test]
    fn test_split_features_nested() {
        let mut features = FeaturesMap::new();
        features.insert("feature1".to_string(), vec!["feature2".to_string()]);
        features.insert("feature2".to_string(), vec![]);
        features.insert("feature3".to_string(), vec!["feature1".to_string()]);

        let (features, features2) = split_features(features);

        assert_compact_debug_snapshot!(features, @r#"{"feature1": ["feature2"], "feature2": [], "feature3": ["feature1"]}"#);
        assert_compact_debug_snapshot!(features2, @"{}");
    }

    #[test]
    fn test_split_features_nested_mixed() {
        let mut features = FeaturesMap::new();
        features.insert("feature1".to_string(), vec!["feature2".to_string()]);
        features.insert("feature2".to_string(), vec!["feature3".to_string()]);
        features.insert("feature3".to_string(), vec!["dep:foo".to_string()]);

        let (features, features2) = split_features(features);

        assert_compact_debug_snapshot!(features, @r#"{"feature1": ["feature2"], "feature2": ["feature3"]}"#);
        assert_compact_debug_snapshot!(features2, @r#"{"feature3": ["dep:foo"]}"#);
    }
}
