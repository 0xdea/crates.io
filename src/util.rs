pub use self::bytes_request::BytesRequest;
pub use self::io_util::{read_fill, read_le_u32};
pub use self::request_helpers::*;

mod bytes_request;
pub mod diesel;
pub mod errors;
mod io_util;
mod request_helpers;
pub mod rfc3339;
pub mod token;
pub mod tracing;
