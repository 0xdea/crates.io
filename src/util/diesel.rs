use diesel::connection::LoadConnection;
use diesel::pg::Pg;
use diesel::result::Error;

pub trait Conn: LoadConnection<Backend = Pg> {}

impl<T> Conn for T where T: LoadConnection<Backend = Pg> {}

pub fn is_read_only_error(error: &Error) -> bool {
    matches!(error, Error::DatabaseError(_, info) if info.message().ends_with("read-only transaction"))
}
