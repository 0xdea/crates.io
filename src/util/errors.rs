use std::io::{IoError, MemReader};
use std::fmt;
use std::fmt::{Show, Formatter, FormatError};
use std::collections::HashMap;

use conduit::Response;
use curl::ErrCode;
use pg::error::PostgresError;
use serialize::json;

pub trait CargoError: Send {
    fn description(&self) -> String;
    fn detail(&self) -> Option<String> { None }
    fn cause<'a>(&'a self) -> Option<&'a CargoError + Send> { None }

    fn to_error<E: FromError<Self>>(self) -> E {
        FromError::from_error(self)
    }

    fn box_error(self) -> Box<CargoError + Send> {
        box self as Box<CargoError + Send>
    }

    fn concrete(&self) -> ConcreteCargoError {
        ConcreteCargoError {
            description: self.description(),
            detail: self.detail(),
            cause: self.cause().map(|c| box c.concrete() as Box<CargoError + Send>),
        }
    }

    fn with_cause<E: CargoError + Send>(self, cause: E) -> Box<CargoError + Send> {
        let mut concrete = self.concrete();
        concrete.cause = Some(cause.box_error());
        box concrete as Box<CargoError + Send>
    }

    fn response(&self) -> Option<Response> { None }
}

pub trait FromError<E> {
    fn from_error(error: E) -> Self;
}

impl<E: CargoError + Send> FromError<E> for Box<CargoError + Send> {
    fn from_error(error: E) -> Box<CargoError + Send> {
        error.box_error()
    }
}

macro_rules! from_error (
    ($ty:ty) => {
        impl FromError<$ty> for $ty {
            fn from_error(error: $ty) -> $ty {
                error
            }
        }
    }
)

impl Show for Box<CargoError + Send> {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        write!(f, "{}", self.description())
    }
}

impl CargoError for Box<CargoError + Send> {
    fn description(&self) -> String {
        (*self).description()
    }

    fn detail(&self) -> Option<String> {
        (*self).detail()
    }

    fn cause<'a>(&'a self) -> Option<&'a CargoError + Send> {
        (*self).cause()
    }

    fn box_error(self) -> Box<CargoError + Send> {
        self
    }
}

pub type CargoResult<T> = Result<T, Box<CargoError + Send>>;

pub trait BoxError<T> {
    fn box_error(self) -> CargoResult<T>;
}

pub trait ChainError<T> {
    fn chain_error<E: CargoError + Send>(self, callback: || -> E) -> CargoResult<T> ;
}

impl<'a, T> ChainError<T> for ||:'a -> CargoResult<T> {
    fn chain_error<E: CargoError + Send>(self, callback: || -> E) -> CargoResult<T> {
        self().map_err(|err| callback().with_cause(err))
    }
}

impl<T, E: CargoError + Send> BoxError<T> for Result<T, E> {
    fn box_error(self) -> CargoResult<T> {
        self.map_err(|err| err.box_error())
    }
}

impl<T, E: CargoError + Send> ChainError<T> for Result<T, E> {
    fn chain_error<E: CargoError + Send>(self, callback: || -> E) -> CargoResult<T>  {
        self.map_err(|err| callback().with_cause(err))
    }
}

impl CargoError for IoError {
    fn description(&self) -> String { self.to_string() }
}

from_error!(IoError)

impl CargoError for FormatError {
    fn description(&self) -> String {
        "formatting failed".to_string()
    }
}

from_error!(FormatError)

impl CargoError for PostgresError {
    fn description(&self) -> String { self.to_string() }
}

from_error!(PostgresError)

impl CargoError for ErrCode {
    fn description(&self) -> String { self.to_string() }
}

from_error!(ErrCode)

impl CargoError for json::DecoderError {
    fn description(&self) -> String { self.to_string() }
}

from_error!(json::DecoderError)

pub struct ConcreteCargoError {
    description: String,
    detail: Option<String>,
    cause: Option<Box<CargoError + Send>>,
}

impl Show for ConcreteCargoError {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        write!(f, "{}", self.description)
    }
}

impl CargoError for ConcreteCargoError {
    fn description(&self) -> String {
        self.description.clone()
    }

    fn detail(&self) -> Option<String> {
        self.detail.clone()
    }

    fn cause<'a>(&'a self) -> Option<&'a CargoError + Send> {
        self.cause.as_ref().map(|c| { let err: &CargoError + Send = *c; err })
    }

    fn with_cause<E: CargoError + Send>(mut self,
                                        err: E) -> Box<CargoError + Send> {
        self.cause = Some(err.box_error());
        box self as Box<CargoError + Send>
    }
}

pub struct NotFound;

impl CargoError for NotFound {
    fn description(&self) -> String { "not found".to_string() }

    fn response(&self) -> Option<Response> {
        Some(Response {
            status: (404, "Not Found"),
            headers: HashMap::new(),
            body: box MemReader::new(Vec::new()),
        })
    }
}

from_error!(NotFound)

pub struct Unauthorized;

impl CargoError for Unauthorized {
    fn description(&self) -> String { "unauthorized".to_string() }

    fn response(&self) -> Option<Response> {
        Some(Response {
            status: (403, "Forbidden"),
            headers: HashMap::new(),
            body: box MemReader::new(Vec::new()),
        })
    }
}

from_error!(Unauthorized)

#[allow(dead_code)]
pub fn internal_error<S1: Str, S2: Str>(error: S1,
                                        detail: S2) -> Box<CargoError + Send> {
    box ConcreteCargoError {
        description: error.as_slice().to_string(),
        detail: Some(detail.as_slice().to_string()),
        cause: None,
    } as Box<CargoError + Send>
}

pub fn internal<S: Show>(error: S) -> Box<CargoError + Send> {
    box ConcreteCargoError {
        description: error.to_string(),
        detail: None,
        cause: None,
    } as Box<CargoError + Send>
}
