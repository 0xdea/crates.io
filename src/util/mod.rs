use std::io::{MemReader, IoError};
use std::io::process::{ProcessOutput, Command};
use std::collections::HashMap;
use std::fmt::Show;
use std::str;

use serialize::{json, Encodable};
use url;

use conduit::{Request, Response, Handler};

pub use self::errors::{CargoError, CargoResult, internal, human, internal_error};
pub use self::errors::{ChainError, BoxError};
pub use self::result::{Require, Wrap};
pub use self::lazy_cell::LazyCell;
pub use self::io::LimitErrorReader;
pub use self::hasher::HashingReader;

pub mod errors;
pub mod result;
mod lazy_cell;
mod io;
mod hasher;

pub trait RequestUtils {
    fn redirect(self, url: String) -> Response;

    fn json<'a, T: Encodable<json::Encoder<'a>, IoError>>(self, t: &T) -> Response;
    fn query(self) -> HashMap<String, String>;
    fn wants_json(self) -> bool;
}

pub fn json_response<'a, T>(t: &T) -> Response
                            where T: Encodable<json::Encoder<'a>, IoError> {
    let s = json::encode(t);
    let mut headers = HashMap::new();
    headers.insert("Content-Type".to_string(),
                   vec!["application/json; charset=utf-8".to_string()]);
    Response {
        status: (200, "OK"),
        headers: headers,
        body: box MemReader::new(s.into_bytes()),
    }
}


impl<'a> RequestUtils for &'a Request + 'a {
    fn json<'a, T: Encodable<json::Encoder<'a>, IoError>>(self, t: &T) -> Response {
        json_response(t)
    }

    fn query(self) -> HashMap<String, String> {
        url::form_urlencoded::parse_str(self.query_string().unwrap_or(""))
            .into_iter().collect()
    }

    fn redirect(self, url: String) -> Response {
        let mut headers = HashMap::new();
        headers.insert("Location".to_string(), vec![url.to_string()]);
        Response {
            status: (302, "Found"),
            headers: headers,
            body: box MemReader::new(Vec::new()),
        }
    }

    fn wants_json(self) -> bool {
        let content = self.headers().find("Accept").unwrap_or(Vec::new());
        content.iter().any(|s| s.contains("json"))
    }
}

pub struct C(pub fn(&mut Request) -> CargoResult<Response>);

impl Handler for C {
    fn call(&self, req: &mut Request) -> Result<Response, Box<Show + 'static>> {
        let C(f) = *self;
        match f(req) {
            Ok(req) => Ok(req),
            Err(e) => {
                match e.response() {
                    Some(response) => Ok(response),
                    None => Err(box e as Box<Show>),
                }
            }
        }
    }
}

pub fn exec(cmd: &Command) -> CargoResult<ProcessOutput> {
    let output = try!(cmd.output().chain_error(|| {
        internal(format!("failed to run command `{}`", cmd))
    }));
    if !output.status.success() {
        let mut desc = String::new();
        if output.output.len() != 0 {
            desc.push_str("--- stdout\n");
            desc.push_str(str::from_utf8(output.output.as_slice()).unwrap());
        }
        if output.error.len() != 0 {
            desc.push_str("--- stderr\n");
            desc.push_str(str::from_utf8(output.error.as_slice()).unwrap());
        }
        Err(internal_error(format!("failed to run command `{}`", cmd), desc))
    } else {
        Ok(output)
    }
}
