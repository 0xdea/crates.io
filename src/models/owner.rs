use diesel::pg::Pg;
use diesel::prelude::*;

use crate::models::{Crate, CrateOwnerInvitation, Team, User};
use crate::schema::crate_owners;
use crates_io_diesel_helpers::pg_enum;

#[derive(Insertable, Associations, Identifiable, Debug, Clone, Copy)]
#[diesel(
    table_name = crate_owners,
    check_for_backend(diesel::pg::Pg),
    primary_key(crate_id, owner_id, owner_kind),
    belongs_to(Crate),
    belongs_to(User, foreign_key = owner_id),
    belongs_to(Team, foreign_key = owner_id),
)]
pub struct CrateOwner {
    pub crate_id: i32,
    pub owner_id: i32,
    pub created_by: i32,
    pub owner_kind: OwnerKind,
    pub email_notifications: bool,
}

type BoxedQuery<'a> = crate_owners::BoxedQuery<'a, Pg, crate_owners::SqlType>;

impl CrateOwner {
    /// Returns a base crate owner query filtered by the owner kind argument. This query also
    /// filters out deleted records.
    pub fn by_owner_kind(kind: OwnerKind) -> BoxedQuery<'static> {
        crate_owners::table
            .filter(crate_owners::deleted.eq(false))
            .filter(crate_owners::owner_kind.eq(kind))
            .into_boxed()
    }

    pub fn from_invite(invite: &CrateOwnerInvitation) -> Self {
        Self {
            crate_id: invite.crate_id,
            owner_id: invite.invited_user_id,
            created_by: invite.invited_by_user_id,
            owner_kind: OwnerKind::User,
            email_notifications: true,
        }
    }
}

pg_enum! {
    pub enum OwnerKind {
        User = 0,
        Team = 1,
    }
}

/// Unifies the notion of a User or a Team.
#[derive(Debug)]
pub enum Owner {
    User(User),
    Team(Team),
}

impl Owner {
    pub fn kind(&self) -> i32 {
        match self {
            Owner::User(_) => OwnerKind::User as i32,
            Owner::Team(_) => OwnerKind::Team as i32,
        }
    }

    pub fn login(&self) -> &str {
        match self {
            Owner::User(user) => &user.gh_login,
            Owner::Team(team) => &team.login,
        }
    }

    pub fn id(&self) -> i32 {
        match self {
            Owner::User(user) => user.id,
            Owner::Team(team) => team.id,
        }
    }
}
