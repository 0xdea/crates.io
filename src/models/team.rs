use bon::Builder;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};

use crates_io_github::{GitHubClient, GitHubError};
use oauth2::AccessToken;

use crate::models::{Crate, CrateOwner, Owner, OwnerKind};
use crate::schema::{crate_owners, teams};

/// For now, just a Github Team. Can be upgraded to other teams
/// later if desirable.
#[derive(Queryable, Identifiable, Serialize, Deserialize, Debug, Selectable)]
pub struct Team {
    /// Unique table id
    pub id: i32,
    /// "github:org:team"
    /// An opaque unique ID, that was at one point parsed out to query Github.
    /// We only query membership with github using the github_id, though.
    /// This is the only name we should ever talk to Cargo about.
    pub login: String,
    /// The GitHub API works on team ID numbers. This can change, if a team
    /// is deleted and then recreated with the same name!!!
    pub github_id: i32,
    /// Sugary goodness
    pub name: Option<String>,
    pub avatar: Option<String>,
    /// The GitHub Organization ID this team sits under
    pub org_id: i32,
}

#[derive(Insertable, AsChangeset, Debug, Builder)]
#[diesel(table_name = teams, check_for_backend(diesel::pg::Pg))]
pub struct NewTeam<'a> {
    pub login: &'a str,
    pub github_id: i32,
    pub name: Option<&'a str>,
    pub avatar: Option<&'a str>,
    pub org_id: i32,
}

impl NewTeam<'_> {
    pub async fn create_or_update(&self, conn: &mut AsyncPgConnection) -> QueryResult<Team> {
        use diesel::insert_into;

        insert_into(teams::table)
            .values(self)
            .on_conflict(teams::github_id)
            .do_update()
            .set(self)
            .get_result(conn)
            .await
    }
}

impl Team {
    /// Phones home to Github to ask if this User is a member of the given team.
    /// Note that we're assuming that the given user is the one interested in
    /// the answer. If this is not the case, then we could accidentally leak
    /// private membership information here.
    pub async fn contains_user(
        &self,
        gh_client: &dyn GitHubClient,
        gh_login: &str,
        token: &AccessToken,
    ) -> Result<bool, GitHubError> {
        team_with_gh_id_contains_user(gh_client, self.org_id, self.github_id, gh_login, token).await
    }

    pub async fn owning(krate: &Crate, conn: &mut AsyncPgConnection) -> QueryResult<Vec<Owner>> {
        let base_query = CrateOwner::belonging_to(krate).filter(crate_owners::deleted.eq(false));
        let teams = base_query
            .inner_join(teams::table)
            .select(Team::as_select())
            .filter(crate_owners::owner_kind.eq(OwnerKind::Team))
            .load(conn)
            .await?
            .into_iter()
            .map(Owner::Team);

        Ok(teams.collect())
    }
}

pub async fn can_add_team(
    gh_client: &dyn GitHubClient,
    org_id: i32,
    team_id: i32,
    gh_login: &str,
    token: &AccessToken,
) -> Result<bool, GitHubError> {
    Ok(
        team_with_gh_id_contains_user(gh_client, org_id, team_id, gh_login, token).await?
            || is_gh_org_owner(gh_client, org_id, gh_login, token).await?,
    )
}

async fn is_gh_org_owner(
    gh_client: &dyn GitHubClient,
    org_id: i32,
    gh_login: &str,
    token: &AccessToken,
) -> Result<bool, GitHubError> {
    let membership = gh_client.org_membership(org_id, gh_login, token).await?;
    Ok(membership.is_some_and(|m| m.is_active_admin()))
}

async fn team_with_gh_id_contains_user(
    gh_client: &dyn GitHubClient,
    github_org_id: i32,
    github_team_id: i32,
    gh_login: &str,
    token: &AccessToken,
) -> Result<bool, GitHubError> {
    // GET /organizations/:org_id/team/:team_id/memberships/:username
    // check that "state": "active"

    let membership = gh_client
        .team_membership(github_org_id, github_team_id, gh_login, token)
        .await?;

    // There is also `state: pending` for which we could possibly give
    // some feedback, but it's not obvious how that should work.
    Ok(membership.is_some_and(|m| m.is_active()))
}
