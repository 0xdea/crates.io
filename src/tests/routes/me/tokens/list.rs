use crate::routes::me::tokens::delete::RevokedResponse;
use crate::util::{RequestHelper, TestApp};
use cargo_registry::models::ApiToken;
use std::collections::HashSet;

#[derive(Deserialize)]
struct DecodableApiToken {
    name: String,
}

#[derive(Deserialize)]
struct ListResponse {
    api_tokens: Vec<DecodableApiToken>,
}

#[test]
fn list_logged_out() {
    let (_, anon) = TestApp::init().empty();
    anon.get("/api/v1/me/tokens").assert_forbidden();
}

#[test]
fn list_with_api_token_is_forbidden() {
    let (_, _, _, token) = TestApp::init().with_token();
    token.get("/api/v1/me/tokens").assert_forbidden();
}

#[test]
fn list_empty() {
    let (_, _, user) = TestApp::init().with_user();
    let json: ListResponse = user.get("/api/v1/me/tokens").good();
    assert_eq!(json.api_tokens.len(), 0);
}

#[test]
fn list_tokens() {
    let (app, _, user) = TestApp::init().with_user();
    let id = user.as_model().id;
    let tokens = app.db(|conn| {
        vec![
            assert_ok!(ApiToken::insert(conn, id, "bar")),
            assert_ok!(ApiToken::insert(conn, id, "baz")),
        ]
    });

    let json: ListResponse = user.get("/api/v1/me/tokens").good();
    assert_eq!(json.api_tokens.len(), tokens.len());
    assert_eq!(
        json.api_tokens
            .into_iter()
            .map(|t| t.name)
            .collect::<HashSet<_>>(),
        tokens
            .into_iter()
            .map(|t| t.model.name)
            .collect::<HashSet<_>>()
    );
}

#[test]
fn list_tokens_exclude_revoked() {
    let (app, _, user) = TestApp::init().with_user();
    let id = user.as_model().id;
    let tokens = app.db(|conn| {
        vec![
            assert_ok!(ApiToken::insert(conn, id, "bar")),
            assert_ok!(ApiToken::insert(conn, id, "baz")),
        ]
    });

    // List tokens expecting them all to be there.
    let json: ListResponse = user.get("/api/v1/me/tokens").good();
    assert_eq!(json.api_tokens.len(), tokens.len());

    // Revoke the first token.
    let _json: RevokedResponse = user
        .delete(&format!("/api/v1/me/tokens/{}", tokens[0].model.id))
        .good();

    // Check that we now have one less token being listed.
    let json: ListResponse = user.get("/api/v1/me/tokens").good();
    assert_eq!(json.api_tokens.len(), tokens.len() - 1);
    assert!(!json
        .api_tokens
        .iter()
        .any(|token| token.name == tokens[0].model.name));
}
