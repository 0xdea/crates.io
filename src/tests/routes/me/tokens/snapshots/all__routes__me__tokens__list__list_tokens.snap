---
source: src/tests/routes/me/tokens/list.rs
expression: response.into_json()
---
api_tokens:
  - crate_scopes: ~
    created_at: "[datetime]"
    endpoint_scopes: ~
    expired_at: ~
    id: "[id]"
    last_used_at: "[datetime]"
    name: bar
  - crate_scopes:
      - serde
      - serde-*
    created_at: "[datetime]"
    endpoint_scopes:
      - publish-update
    expired_at: ~
    id: "[id]"
    last_used_at: "[datetime]"
    name: baz

